src/main/cpp/audioengine.cpp

```diff
@@ -419,12 +420,19 @@ namespace AudioEngine
             }
         }
 
-        // write the synthesized output into the audio driver
+        // thread has been stopped during operations above ? exit as writing the
+        // the output into the audio hardware will lock execution until the next buffer
+        // is enqueued (additionally, we prevent writing to device storage when recording/bouncing)
+
+        if ( thread == 0 )
+            return false;
+
```

Comment:
Previously rendered audio was written into the audio hardware (and onto device storage when recording/bouncing) before completing the render method. We now allow an early exit. If after rendering it turns out the thread has been stopped on the outside (because the engine is paused()/destroyed(), we return here.

The reason being that writing samples into the audio hardware is a blocking operation which stalls this threads execution and thus prevents us from immediately freeing the resources allocated by the audio drivers.